Support complex polygons for geofences

ascandella opened the pull request:
cc @countrodrigo @jedhorne @kmnovak

Commit:
Support complex polygon in Geofence class

This commit adds support for "interior rings" aka "holes" inside
polygons as per the GeoJSON spec:

  http://www.geojson.org/geojson-spec.html#polygon

See discussion thread here:

  http://lists.geojson.org/pipermail/geojson-geojson.org/2013-November/000863.html

Maniphest task: https://code.uber.com/T1955

dfellis commented on lib/geofence.js:
```diff
@@ -5,7 +5,18 @@ var Geofence = function(vertices, granularity) {
+        throw new Error("Vertices must contain at least one point");
+    }
+
+    // Support complexy polygons a la GeoJSON with inner rings removed
```
Typo. complexy => complex

dfellis commented on lib/geofence.js:
```diff
@@ -5,7 +5,18 @@ var Geofence = function(vertices, granularity) {
+    }
+
+    // Support complexy polygons a la GeoJSON with inner rings removed
+    if (Array.isArray(vertices[0][0])) {
```
I'm confused why the same couldn't be accomplished with a combination of blackout and whiteout geofences, rather than complicating the intersection algorithm?

ascandella replied:
Please see the discussion here: https://code.uber.com/T1955

Basically, it makes more sense to treat it as a single entity (as geojson does) rather than re-computing every time.

dfellis replied:
Can I assume the geojson is defining something like this?

![polygons](https://f.cloud.github.com/assets/765551/1518233/9605d332-4b33-11e3-8b10-89d08b973641.png)

That is, polygons simply invert the in/out status of the area they contain?

If so, the `holes` nomenclature is a bit misleading and the algorithm below will return `false` in some cases it shouldn't. (And you could also reduce the algorithm to simply iterate through all of the polygons and xor each result on the previous.)

ascandella replied:
Not quite; the inner list of rings basically only represents holes
http://www.geojson.org/geojson-spec.html#id18
the "coordinates" member must be an array of LinearRing coordinate arrays. For Polygons with multiple rings, the first must be the exterior ring and any others must be interior rings or holes.

So xoring is incorrect and I believe this formula does the right thing. I'm of course open to suggestions but I believe this properly supports geojson polygons.

dfellis replied:
Hmm... That spec doesn't define what an "interior ring" is or what a "hole" is. In fact, [grepping through the entire specification](https://github.com/GeoJSONWG/geojson-spec/blob/master/geojson-spec-draft.rst) yields no definition of "interior ring" at all.

Is there a reference implementation for GeoJSON somewhere that can clarify things? I still think the xoring approach would yield a more useful behavior that can be directly created in vector drawing tools like Inkscape (it's the object "exclusion" merge type) and would yield a very simple merge algorithm that doesn't have to treat inner and outer geometries differently.

I wouldn't think they'd ignore the past 30-some years of graphics research on this. (I know, I know, it's the JS community, but still.)

ascandella replied:
I havent found a reference implementation yet. I'll do some more searching when I get back to SF so I can be sure I'm implementing this the "right" way.

dfellis replied:
[Here's the answer](http://lists.geojson.org/pipermail/geojson-geojson.org/2013-November/000863.html)

tl;dr: There is no "right" way. If it passes [geojsonlint](http://geojsonlint.com/) it's valid GeoJSON, and that's just confirming JSON structure, not that the interior rings are actually inside of the exterior ring or how to interpret what the interior rings _are_, or anything of that nature.

Personally, I'd prefer we treat it as exclusions, since the "exterior" ring doesn't have to be totally encompassing the whole area, especially since I can see some exceptions to the rule that this would allow (willing to discuss this 1:1 with you and Jed so you know what I mean, but don't want to put it in a public repo).

dfellis commented on lib/geofence.js:
```diff
@@ -16,8 +27,9 @@ var Geofence = function(vertices, granularity) {
     // Close the geofence if necessary
-    if (vertices[0][0] !== vertices[vertices.length - 1][0] || vertices[0][1] !== vertices[vertices.length - 1][1]) {
-        vertices.push(vertices[0]);
+    if (this.vertices[0][0] !== this.vertices[vertices.length - 1][0] ||
```
Shouldn't this check be done for all of the `holes`, too?

ascandella replied:
I don't think it's necessary... this check is for non-compliant geojson, whereas the support that's being added is for well-vetted objects. I maintain the legacy check for compatibility but do not complicate the logic going forward just for invalid data.

dfellis replied:
But in-n-out is _not_ doing any vetting, itself. This either needs to be well documented, or perhaps throw an error in case the `holes` aren't well-defined?

ascandella replied:
I don't feel strongly one way or another. You're right that this behavior should be more transparent though. I'm inclined to throw an error. What do you think?

dfellis replied:
`throw` an error. We can try-catch creation of these new style GeoJSON polygons and log the error so upstream can know that their output has a bug. :)

ascandella replied:
Actually, I think all of this is unnecessary; I removed the 'closing' logic and all of the tests still passed. I think because we defer to point-in-polygon, we can let it handle the closing for us.

dfellis replied:
I'd [take a look at the point-in-polygon source](https://github.com/substack/point-in-polygon/blob/master/index.js) as it does no such thing, which actually implies that the test suite in in-n-out needs improvement.

dfellis commented on lib/geofence.js:
```diff
@@ -79,39 +100,41 @@ Geofence.prototype = {
-            for (var tileY = minTileY; tileY <= maxTileY; tileY++) {
-                tileHash = (tileY - minTileY) * this.granularity + (tileX - minTileX);
+    setExclusionTiles: function(vertices, inclusive) {
+        var bBoxPoly,
```
Separate var blocks, please.

ascandella replied:
Sure thing.

truncs commented on lib/geofence.js:
```diff
@@ -52,7 +62,18 @@ Geofence.prototype = {
+            // If we do have holes cut out, and the point falls within the outer
+            // ring, ensure no inner rings exclude this point
+            for (var i = 0, hole; hole = this.holes[i]; i++) {
+                if (utils.pointInPolygon(point, hole)) {
```
What happens when the point is on the boundary of the hole ? Also totally unrelated, but the polygons that are constructed using gps coordinates might not work correctly over a large surface area because the polygon is not curved along the surface of the earth.

dfellis replied:
Well, I don't think we're too concerned about that sort of boundary condition? Conceivably they'd flutter between the two sides on each ping based on the GPS error. As for what the code will actually do, being on the boundary means you're "in" that boundary, so it'd probably put them in the hole.

As for point 2, when ops creates their polygons they're looking at such a lat/lon projection in the first place and they're probably transcribing legal boundary areas directly in that same projection with Google Maps (or city bounds maps, etc) so it should still be accurate (because someone else did the projection work for them).

truncs replied:
Interesting thoughts @dfellis, yes I didn't think about that, google maps uses kind of mercator projection. Also I went and had a peek in point-in-polygon library (didn't look at the code carefully though) which uses ray tracing and should work on complex polygons as well directly ( if we give a set of x-y coordinates for the polygon) but this way its simpler.

countrodrigo replied:
This is a performance issue. The magic behind in-n-out is the cached griding for each geofence ring, resulting in only needing to do an O(N) ray trace check in the worse case of intersecting tiles and O(1) otherwise. We need to compute a cached grid for each hole ring and use that instead for the hole checks.

ascandella replied:
Did you actually read my implementation?

You're looking at the 'x' block, which only applies if the point lies on the boundary. 

Trust me, I understand the point of the library, and for points that are clearly within the inner ring(s) at the given resolution, we short-circuit:

https://github.com/uber/in-n-out/pull/7/files#diff-068591b5cb8abf665e6104587d821d57R139

dfellis commented:
I'd still like this to switch to an xor-ing mode of secondary polygons rather than simply negating. It'd meet all of Jed's needs, simplify the algorithm, and be a more general-purpose solution that could be applied in other areas we use in-n-out.

countrodrigo commented on lib/geofence.js:
```diff
@@ -52,7 +68,18 @@ Geofence.prototype = {
             // this.testsIntersecting++;
-            return utils.pointInPolygon(point, this.vertices);
+            var inside = utils.pointInPolygon(point, this.vertices);
+            if (!this.holes || !inside) {
```
Let's check for !inside first, since that would make us return false right away without having to check for holes.

ascandella replied:
Fixed, although it's just a property lookup so is not really very expensive.

ascandella commented:
@dfellis I'm still on the fence about xor-ing: on the one hand, it might make the algorithm simpler (although I think the algorithm is pretty dead-simple as presented): on the other hand, I'm not sure it's the desired behavior. I've updated the ticket on Phabricator to reflect the current state of things.

Other than that, I've updated edge-case handling (I decided to close the loop in all cases), PR message, commit history, version bumping, and test coverage. Please let me know if you have any more feedback!

Thanks for the thorough review :)

countrodrigo commented on package.json:
```diff
@@ -8,8 +8,8 @@
   },
   "author": "Alain Rodriguez <eagle5command@gmail.com>",
 
```
Add yourself to the contributor list

countrodrigo commented:
Also, look at this PR: https://github.com/uber/in-n-out/pull/4. It accomplishes something very similar but in a performant way.

ascandella commented:
@countrodrigo why are you just now suggesting using an alternative pull request? This has been open for over a week, you were included on the original ticket, and we've now sunk many man hours into this. If you really wanted this done a different way, why didn't you speak up earlier?

countrodrigo commented:
@sectioneight You are correct. My differ was only showing me the intersect portion. Sorry for the confusion.

countrodrigo commented:
This is good to go as is. Merging.
